File: mcp/tools/modelcontextprotocol-inspector/core/json/xMcpHeader.ts
/**
 * SEP-2243 `x-mcp-header` annotation tooling.
 *
 * A modern (≥2026-07-28) Streamable HTTP server may annotate a tool
 * `inputSchema` property with `x-mcp-header: "{Name}"`; a conforming client then
 * mirrors that argument's value into an `Mcp-Param-{Name}` HTTP header on the
 * `tools/call`. The spec places strict constraints on which properties may carry
 * the annotation, and — crucially for a debugging tool — makes a *violating*
 * annotation invalidate the WHOLE tool: a Streamable HTTP client MUST drop such
 * a tool from `tools/list`.
 *
 * The client SDK enforces that exclusion internally (its `listTools()` silently
 * filters invalid tools), but it does not surface *which* tools were dropped or
 * *why*. This module re-implements the SDK's scan so the Inspector can show
 * excluded tools with their reason, and indicate which args mirror to headers on
 * the tools it keeps. It is a faithful port of the SDK's
 * `scanXMcpHeaderDeclarations` (the helper is not part of the SDK's public
 * surface), kept pure and fully unit-testable — no rendering, no I/O.
 */

import type { Tool } from "@modelcontextprotocol/client";

/** The schema-extension property name a tool's `inputSchema` carries. */
export const X_MCP_HEADER_KEY = "x-mcp-header";

/** The fixed prefix every mirrored custom-parameter header carries. */
export const MCP_PARAM_HEADER_PREFIX = "Mcp-Param-";

/**
 * RFC 9110 §5.1 `token` syntax (`1*tchar`). Rejects empty, space, control
 * characters (including CR/LF), and the listed HTTP delimiters.
 */
const RFC9110_TOKEN = /^[!#$%&'*+\-.^_`|~0-9A-Za-z]+$/;

/**
 * JSON Schema `type` values the spec admits on an `x-mcp-header` property.
 *
 * The spec text names `integer`, `string`, `boolean` and explicitly excludes
 * `number`. The published conformance referee at the pinned release ships its
 * `http-custom-headers` scenario with `type: "number"` `x-mcp-header` params and
 * expects the client to mirror them, so the SDK accepts `number` for the
 * conformance gate; this port matches the SDK so exclusions agree exactly.
 * Everything else (`object`, `array`, `null`, absent) is rejected.
 */
const PERMITTED_X_MCP_HEADER_TYPES: ReadonlySet<string> = new Set([
  "string",
  "integer",
  "boolean",
  "number",
]);

/**
 * JSON Schema keywords whose subschemas the static-reachability constraint
 * excludes from the `properties`-only chain. An `x-mcp-header` found under any
 * of these invalidates the tool definition.
 */
const NON_REACHABLE_SUBSCHEMA_KEYWORDS = [
  "items",
  "prefixItems",
  "contains",
  "additionalProperties",
  "unevaluatedProperties",
  "unevaluatedItems",
  "propertyNames",
  "patternProperties",
  "dependentSchemas",
  "oneOf",
  "anyOf",
  "allOf",
  "not",
  "if",
  "then",
  "else",
  "$defs",
  "definitions",
] as const;

/**
 * Subschema-carrying keywords whose value is a `name → subschema` object (not a
 * single subschema or array of subschemas). The visit branches over
 * `Object.values()` for these.
 */
const OBJECT_VALUED_SUBSCHEMA_KEYWORDS: ReadonlySet<string> = new Set([
  "patternProperties",
  "dependentSchemas",
  "$defs",
  "definitions",
]);

/** One validated `x-mcp-header` declaration found on a tool's input schema. */
export interface XMcpHeaderDeclaration {
  /** The chain of `properties` keys locating the annotated property. */
  path: string[];
  /** The declared header suffix — the `{Name}` in `Mcp-Param-{Name}`. */
  headerName: string;
  /** The property's JSON Schema `type` (a permitted primitive). */
  type: string;
}

/** The result of scanning a tool's input schema for `x-mcp-header` usage. */
export type XMcpHeaderScan =
  | { valid: true; declarations: XMcpHeaderDeclaration[] }
  | { valid: false; reason: string };

function pathName(path: string[]): string {
  return path.length === 0 ? "<root>" : path.join(".");
}

function isRecord(node: unknown): node is Record<string, unknown> {
  return node !== null && typeof node === "object";
}

/**
 * Scan a tool's `inputSchema` for `x-mcp-header` declarations and validate every
 * constraint the spec places on them. Returns the collected declarations
 * (possibly empty) on success, or the first violated constraint's reason.
 *
 * The walk descends through `properties` at any depth (the spec's "any nesting
 * depth" clause). The static-reachability MUST is enforced structurally: every
 * position the chain MUST NOT pass through (`items`/`additionalProperties`,
 * `oneOf`/`anyOf`/`allOf`/`not`, `if`/`then`/`else`, and `$defs`/`definitions`
 * bodies) is visited too, and an `x-mcp-header` found anywhere off the
 * `properties` chain invalidates the schema — "an annotation anywhere else makes
 * the tool definition invalid". `$ref` is never followed: a property reachable
 * only through a `$ref` is therefore correctly treated as non-statically-
 * reachable (its annotation, if any, lives in the unreachable `$defs` body).
 */
export function scanXMcpHeaderDeclarations(
  inputSchema: unknown,
): XMcpHeaderScan {
  const declarations: XMcpHeaderDeclaration[] = [];
  const seenLower = new Map<string, string>();

  const visit = (
    node: unknown,
    path: string[],
    reachable: boolean,
  ): string | undefined => {
    if (!isRecord(node)) return undefined;
    const schema = node;

    if (X_MCP_HEADER_KEY in schema) {
      if (!reachable || path.length === 0) {
        return `${pathName(path)}: x-mcp-header is only permitted on properties statically reachable via a chain of 'properties' keys (not under items, additionalProperties, oneOf/anyOf/allOf/not, if/then/else, or $ref)`;
      }
      const raw = schema[X_MCP_HEADER_KEY];
      if (typeof raw !== "string" || raw.length === 0) {
        return `${pathName(path)}: x-mcp-header MUST be a non-empty string`;
      }
      if (!RFC9110_TOKEN.test(raw)) {
        return `${pathName(path)}: x-mcp-header '${raw}' is not a valid RFC 9110 token (no spaces, control characters or HTTP delimiters)`;
      }
      const type = typeof schema.type === "string" ? schema.type : undefined;
      if (type === undefined || !PERMITTED_X_MCP_HEADER_TYPES.has(type)) {
        return `${pathName(path)}: x-mcp-header is only permitted on primitive-typed properties (string, integer, boolean); got ${type ?? "<none>"}`;
      }
      const lower = raw.toLowerCase();
      const prior = seenLower.get(lower);
      if (prior !== undefined) {
        return `x-mcp-header '${raw}' is not case-insensitively unique (also declared as '${prior}')`;
      }
      seenLower.set(lower, raw);
      declarations.push({ path, headerName: raw, type });
    }

    const properties = schema.properties;
    if (isRecord(properties)) {
      for (const [key, child] of Object.entries(properties)) {
        const fault = visit(child, [...path, key], reachable);
        if (fault !== undefined) return fault;
      }
    }

    for (const k of NON_REACHABLE_SUBSCHEMA_KEYWORDS) {
      const sub = schema[k];
      if (sub === undefined) continue;
      const branches = Array.isArray(sub)
        ? sub
        : isRecord(sub) && OBJECT_VALUED_SUBSCHEMA_KEYWORDS.has(k)
          ? Object.values(sub)
          : [sub];
      for (const branch of branches) {
        const fault = visit(branch, [...path, `<${k}>`], false);
        if (fault !== undefined) return fault;
      }
    }

    return undefined;
  };

  const fault = visit(inputSchema, [], true);
  return fault === undefined
    ? { valid: true, declarations }
    : { valid: false, reason: fault };
}

/**
 * The `=?base64?…?=` sentinel wrapping a value that cannot be sent as a plain
 * ASCII HTTP field value (SEP-2243 value-encoding rules).
 */
const BASE64_SENTINEL_PREFIX = "=?base64?";
const BASE64_SENTINEL_SUFFIX = "?=";

/**
 * Convert a primitive argument value to its string form per the spec's
 * type-conversion rules: strings pass through, booleans become lowercase
 * `'true'`/`'false'`, integers/numbers become their decimal string. Non-finite
 * numbers and integers outside the safe range are refused (returns `undefined`,
 * meaning "do not emit a header for this value"). Anything non-primitive
 * (object/array/null/undefined) also yields `undefined`.
 */
function mcpParamPrimitiveToString(value: unknown): string | undefined {
  if (typeof value === "string") return value;
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) return undefined;
    if (Number.isInteger(value) && !Number.isSafeInteger(value)) {
      return undefined;
    }
    return String(value);
  }
  return undefined;
}

/**
 * `true` when `s` cannot be safely represented as a plain ASCII HTTP field
 * value (RFC 9110 §5.5): it is empty, contains a byte outside `0x20–0x7E`/`0x09`,
 * has leading/trailing whitespace (which field parsing strips), or already
 * matches the Base64 sentinel pattern (the spec's "to avoid ambiguity" rule).
 */
function needsBase64(s: string): boolean {
  if (s.length === 0) return true;
  if (
    s.startsWith(BASE64_SENTINEL_PREFIX) &&
    s.endsWith(BASE64_SENTINEL_SUFFIX)
  ) {
    return true;
  }
  if (s !== s.trim()) return true;
  for (let i = 0; i < s.length; i++) {
    // Non-null: `i` is always in bounds, so `codePointAt` returns a number.
    const c = s.codePointAt(i)!;
    if (c === 9 || (c >= 32 && c <= 126)) continue;
    return true;
  }
  return false;
}

function utf8ToBase64(s: string): string {
  const bytes = new TextEncoder().encode(s);
  let bin = "";
  for (const b of bytes) bin += String.fromCodePoint(b);
  return btoa(bin);
}

/**
 * Encode a string value as an HTTP field value per SEP-2243: a value that is
 * already a safe plain-ASCII field value passes through unchanged; anything
 * else is wrapped as `=?base64?{b64-of-utf8}?=`.
 */
function encodeMcpParamValue(value: string): string {
  return needsBase64(value)
    ? `${BASE64_SENTINEL_PREFIX}${utf8ToBase64(value)}${BASE64_SENTINEL_SUFFIX}`
    : value;
}

function valueAtPath(root: unknown, path: string[]): unknown {
  let node: unknown = root;
  for (const key of path) {
    if (node === null || typeof node !== "object") return undefined;
    node = (node as Record<string, unknown>)[key];
  }
  return node;
}

/**
 * Build the `Mcp-Param-{Name}` headers for one `tools/call` from validated
 * `x-mcp-header` declarations and the call's `arguments`. A declaration whose
 * value is `null` or absent is omitted (the spec's "client MUST omit the header"
 * rows); a value that is not a primitive of the declared kind is omitted rather
 * than emitted malformed. Faithful port of the SDK's internal helper (not part
 * of its public surface), so the headers match what a conforming client sends.
 */
export function buildMcpParamHeaders(
  declarations: XMcpHeaderDeclaration[],
  args: Record<string, unknown>,
): Record<string, string> {
  const out: Record<string, string> = {};
  for (const decl of declarations) {
    const raw = valueAtPath(args, decl.path);
    if (raw === undefined || raw === null) continue;
    const stringValue = mcpParamPrimitiveToString(raw);
    if (stringValue === undefined) continue;
    out[`${MCP_PARAM_HEADER_PREFIX}${decl.headerName}`] =
      encodeMcpParamValue(stringValue);
  }
  return out;
}

/**
 * SEP-2243 `Mcp-Param-*` headers a `tools/call` must carry for a given tool and
 * arguments. Returns `{}` when the tool declares no `x-mcp-header`, when its
 * annotations are invalid (such a tool is excluded from `tools/list`), or when
 * no declared argument has a mirrorable value. Callers attach the result to the
 * `tools/call` request headers on a modern connection.
 */
export function mcpParamHeadersForTool(
  tool: Tool,
  args: Record<string, unknown>,
): Record<string, string> {
  const scan = scanXMcpHeaderDeclarations(tool.inputSchema);
  if (!scan.valid || scan.declarations.length === 0) return {};
  return buildMcpParamHeaders(scan.declarations, args);
}

/** A tool the Inspector keeps, paired with its mirrored-header declarations. */
export interface MirroredHeaderParam {
  /** Dot-joined property path (e.g. `region` or `filter.city`). */
  path: string;
  /** The full header a conforming client sends: `Mcp-Param-{Name}`. */
  header: string;
  /** The declared header suffix. */
  headerName: string;
  /** The property's JSON Schema primitive type. */
  type: string;
}

/**
 * The mirrored-header params for a tool the Inspector kept (its annotations are
 * all valid). Returns `[]` when the tool declares no `x-mcp-header`, and — since
 * a caller only reaches here for *kept* tools — also `[]` for the (unreachable
 * for kept tools) invalid case. Each entry names the arg and the
 * `Mcp-Param-{Name}` header its value mirrors to on a `tools/call`.
 */
export function getMirroredHeaderParams(tool: Tool): MirroredHeaderParam[] {
  const scan = scanXMcpHeaderDeclarations(tool.inputSchema);
  if (!scan.valid) return [];
  return scan.declarations.map((d) => ({
    path: d.path.join("."),
    header: `${MCP_PARAM_HEADER_PREFIX}${d.headerName}`,
    headerName: d.headerName,
    type: d.type,
  }));
}

File: mcp/security/toolhive/pkg/mcp/xmcpheader.go
// SPDX-FileCopyrightText: Copyright 2026 Stacklok, Inc.
// SPDX-License-Identifier: Apache-2.0

package mcp

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"slices"
	"strconv"
	"strings"
)

// SEP-2243 ("HTTP standardization") lets an MCP server mark individual tool
// parameters for mirroring into HTTP request headers, via an x-mcp-header
// annotation inside the parameter's schema in the tool's inputSchema. A server
// MAY use the annotation; a client MUST honour it, sending each designated
// parameter's value as the header Mcp-Param-{name} on the tools/call request. A
// server that designated a parameter and did not receive its header rejects the
// call with -32020.
//
// This file owns the annotation's vocabulary and its validation. It is
// deliberately the single place that reads x-mcp-header, so the two consumers --
// ingestion-time validation of a backend's advertised tools, and call-time
// derivation of the outgoing headers -- cannot drift in their reading of the
// constraints. (Compare the "third independent copy" problem called out for the
// reserved _meta keys in #5986.)
const (
	// XMCPHeaderAnnotation is the JSON Schema extension key a server sets on a
	// tool parameter to designate it for header mirroring. Its value is the
	// header's name suffix, NOT the full header name.
	XMCPHeaderAnnotation = "x-mcp-header"

	// ParamHeaderPrefix prefixes every mirrored parameter header: an annotation
	// of "Region" is sent as "Mcp-Param-Region".
	ParamHeaderPrefix = "Mcp-Param-"
)

// maxSchemaDepth bounds the inputSchema traversal. Annotations are legal at any
// nesting depth, so the walk has to recurse, and the schema is attacker-supplied
// from vMCP's perspective (a backend can advertise any tool list it likes). A
// depth cap keeps a hostile or accidentally-recursive schema from exhausting the
// stack. 64 is far past any hand-written tool schema.
const maxSchemaDepth = 64

// ErrSchemaTooDeep is returned when an inputSchema nests past maxSchemaDepth.
// It is deliberately distinguishable from a constraint violation: the schema may
// be perfectly valid and merely beyond what this walk will inspect, so a caller
// that wants to fail open on depth alone can single it out.
var ErrSchemaTooDeep = errors.New("tool inputSchema exceeds maximum inspection depth")

// ParamHeader is one x-mcp-header annotation found in a tool's inputSchema.
type ParamHeader struct {
	// Path is the property path from the root of inputSchema to the annotated
	// parameter, e.g. ["filter", "region"] for a nested property. Length is
	// always >= 1.
	Path []string

	// Name is the annotation's value -- the suffix appended to
	// ParamHeaderPrefix, not the full header name. Use HeaderName for that.
	Name string

	// Type is the annotated parameter's declared JSON Schema type, guaranteed by
	// validation to be one of "string", "integer", or "boolean".
	Type string
}

// HeaderName is the full HTTP header name this annotation mirrors into.
func (p ParamHeader) HeaderName() string {
	return ParamHeaderPrefix + p.Name
}

// ParamHeaders walks a tool's inputSchema and returns every x-mcp-header
// annotation it declares, in a deterministic order (by path). A schema with no
// annotations -- the overwhelmingly common case -- returns nil, nil.
//
// It returns an error when the schema violates any of SEP-2243's constraints on
// the annotation:
//
//   - the value must be a non-empty string;
//   - it must be a valid HTTP field-name token (RFC 9110 tchar), which also
//     excludes control characters and whitespace;
//   - it must be unique, case-insensitively, within the whole inputSchema
//     (HTTP field names are case-insensitive, so two spellings would collide on
//     the wire);
//   - it may only annotate a parameter whose declared type is "string",
//     "integer", or "boolean" -- notably NOT "number", which SEP-2243 excludes
//     because a float has no canonical wire spelling.
//
// A missing or non-string "type" on an annotated parameter is treated as a
// violation. SEP-2243 permits the annotation only on primitive parameters, and
// an undeclared type is not a declared primitive; equally, mirroring cannot
// serialize a value whose type it does not know. This is the strict reading:
// it rejects the tool rather than guessing a spelling for the header value.
//
// Traversal covers "properties", array "items", and the "oneOf"/"anyOf"/"allOf"
// combinators, so an annotation is found wherever SEP-2243 allows one. Nesting
// past maxSchemaDepth yields ErrSchemaTooDeep.
func ParamHeaders(schema map[string]any) ([]ParamHeader, error) {
	if len(schema) == 0 {
		return nil, nil
	}
	// seen maps the case-folded annotation value to the path that first claimed
	// it, so a duplicate can name both colliding parameters in its error.
	seen := map[string]string{}
	var found []ParamHeader
	if err := walkSchema(schema, nil, 0, seen, &found); err != nil {
		return nil, err
	}
	if len(found) == 0 {
		return nil, nil
	}
	// Deterministic order so callers (and tests) see a stable sequence
	// regardless of Go's map iteration order.
	slices.SortFunc(found, func(a, b ParamHeader) int {
		return strings.Compare(strings.Join(a.Path, "."), strings.Join(b.Path, "."))
	})
	return found, nil
}

// ValidateParamHeaders reports whether a tool's inputSchema declares only
// SEP-2243-conformant x-mcp-header annotations. It is ParamHeaders with the
// annotations discarded, for the ingestion-time check where only the verdict
// matters.
func ValidateParamHeaders(schema map[string]any) error {
	_, err := ParamHeaders(schema)
	return err
}

// walkSchema recurses one schema node, appending any annotation it finds to
// found. path is the property path to this node ("" at the root, which cannot
// itself be annotated -- an annotation designates a parameter, and the root is
// the parameter object).
func walkSchema(node map[string]any, path []string, depth int, seen map[string]string, found *[]ParamHeader) error {
	if depth > maxSchemaDepth {
		return fmt.Errorf("%w (%d)", ErrSchemaTooDeep, maxSchemaDepth)
	}

	// An annotation on the root node designates no parameter, so it is ignored
	// rather than treated as an error: len(path) == 0 only at the root.
	if len(path) > 0 {
		if raw, ok := node[XMCPHeaderAnnotation]; ok {
			hdr, err := parseAnnotation(raw, node, path, seen)
			if err != nil {
				return err
			}
			*found = append(*found, hdr)
		}
	}

	return walkChildren(node, path, depth, seen, found)
}

// walkChildren recurses into every sub-schema of node that SEP-2243 allows an
// annotation to appear in: object properties, array element schemas, and the
// oneOf/anyOf/allOf combinator branches. Split from walkSchema to keep each
// within the cyclomatic limit.
func walkChildren(node map[string]any, path []string, depth int, seen map[string]string, found *[]ParamHeader) error {
	if props, ok := node["properties"].(map[string]any); ok {
		for name, sub := range props {
			subSchema, ok := sub.(map[string]any)
			if !ok {
				continue // not a schema object; nothing to inspect
			}
			if err := walkSchema(subSchema, childPath(path, name), depth+1, seen, found); err != nil {
				return err
			}
		}
	}

	// Array element schemas: "items" is a schema object in the JSON Schema
	// dialect MCP uses. The element carries no property name of its own, so it
	// inherits the array's path with an "[]" marker for legibility in errors.
	if items, ok := node["items"].(map[string]any); ok {
		if err := walkSchema(items, childPath(path, "[]"), depth+1, seen, found); err != nil {
			return err
		}
	}

	return walkCombinators(node, path, depth, seen, found)
}

// walkCombinators recurses into the oneOf/anyOf/allOf branches of node. These
// carry real schemas in this repo's backend tool sets (#5976 fixed ingestion
// dropping them), so an annotation inside a branch must be validated like any
// other.
func walkCombinators(
	node map[string]any, path []string, depth int, seen map[string]string, found *[]ParamHeader,
) error {
	for _, combinator := range []string{"oneOf", "anyOf", "allOf"} {
		branches, ok := node[combinator].([]any)
		if !ok {
			continue
		}
		for i, branch := range branches {
			branchSchema, ok := branch.(map[string]any)
			if !ok {
				continue
			}
			marker := fmt.Sprintf("%s[%d]", combinator, i)
			if err := walkSchema(branchSchema, childPath(path, marker), depth+1, seen, found); err != nil {
				return err
			}
		}
	}
	return nil
}

// maxSafeInteger is JavaScript's Number.MAX_SAFE_INTEGER (2^53 - 1). SEP-2243
// requires a mirrored integer to sit within this range, since a peer that parses
// the header with JSON-number semantics could not round-trip a larger value.
const maxSafeInteger = 1<<53 - 1

// ErrUnmirrorableValue is returned when a designated parameter's value cannot be
// mirrored into a header: a control character (header injection), a non-integral
// or out-of-safe-range integer, or a value whose type contradicts the schema.
var ErrUnmirrorableValue = errors.New("parameter value cannot be mirrored into an HTTP header")

// MirrorParamHeaders derives the Mcp-Param-* headers to send with a tools/call,
// given the tool's x-mcp-header annotations (from ParamHeaders) and the call's
// arguments. The result maps full header names to values, ready to set on the
// request; it is nil when nothing is to be mirrored.
//
// A designated parameter that is absent from args contributes no header. That is
// deliberate rather than an error: an optional parameter the caller did not
// supply has no value to mirror, and SEP-2243's -32020 covers the server's view
// of a genuinely missing designated value.
//
// Annotations reached through an array element ("[]" in the path) are skipped: an
// array holds many elements and a header holds one value, so there is no
// well-defined single value to send. Combinator segments (oneOf/anyOf/allOf) are
// dropped when resolving, because they are structural to the schema and absent
// from the arguments the schema describes.
//
// It returns an error wrapping ErrUnmirrorableValue when a present value cannot
// be safely rendered. Values originate from the caller (ultimately a model), so
// they are untrusted: a CR, LF, or NUL in a string would let a caller forge
// additional headers on vMCP's outgoing request, and is refused rather than
// silently stripped.
func MirrorParamHeaders(headers []ParamHeader, args map[string]any) (map[string]string, error) {
	if len(headers) == 0 || len(args) == 0 {
		return nil, nil
	}
	var out map[string]string
	for _, h := range headers {
		if slices.Contains(h.Path, "[]") {
			continue
		}
		value, ok := resolveArg(args, h.Path)
		if !ok {
			continue
		}
		rendered, err := renderHeaderValue(h, value)
		if err != nil {
			return nil, err
		}
		if out == nil {
			out = map[string]string{}
		}
		out[h.HeaderName()] = rendered
	}
	return out, nil
}

// ParamHeadersForSchema is ParamHeaders followed by MirrorParamHeaders: it takes
// a tool's inputSchema and a call's arguments and returns the Mcp-Param-* headers
// to send. It exists so the several call sites that mirror headers share one
// reading of the two-step dance rather than each re-deriving it.
//
// A schema error is returned unwrapped from ParamHeaders; a value error wraps
// ErrUnmirrorableValue. Callers distinguish them because the first indicts the
// backend's tool definition and the second the caller's arguments.
func ParamHeadersForSchema(schema map[string]any, args map[string]any) (map[string]string, error) {
	annotations, err := ParamHeaders(schema)
	if err != nil {
		return nil, err
	}
	if len(annotations) == 0 {
		return nil, nil
	}
	return MirrorParamHeaders(annotations, args)
}

// resolveArg walks path through args and returns the value at its end. Segments
// naming a combinator are skipped: they exist in the schema's structure, not in
// the data it describes. A missing or non-object intermediate reports false.
func resolveArg(args map[string]any, path []string) (any, bool) {
	current := any(args)
	for _, seg := range path {
		if isCombinatorSegment(seg) {
			continue
		}
		obj, ok := current.(map[string]any)
		if !ok {
			return nil, false
		}
		current, ok = obj[seg]
		if !ok {
			return nil, false
		}
	}
	return current, true
}

// isCombinatorSegment reports whether a path segment is a combinator marker
// emitted by walkCombinators (e.g. "oneOf[0]") rather than a property name.
func isCombinatorSegment(seg string) bool {
	for _, combinator := range []string{"oneOf", "anyOf", "allOf"} {
		if strings.HasPrefix(seg, combinator+"[") && strings.HasSuffix(seg, "]") {
			return true
		}
	}
	return false
}

// renderHeaderValue converts a designated parameter's value to its header
// spelling, enforcing the schema's declared type and SEP-2243's integer range.
func renderHeaderValue(h ParamHeader, value any) (string, error) {
	where := strings.Join(h.Path, ".")
	switch h.Type {
	case "string":
		s, ok := value.(string)
		if !ok {
			return "", fmt.Errorf("%w: parameter %q is declared string but got %T", ErrUnmirrorableValue, where, value)
		}
		if bad, invalid := firstControlChar(s); invalid {
			return "", fmt.Errorf(
				"%w: parameter %q contains a control character (%q)", ErrUnmirrorableValue, where, bad)
		}
		return s, nil
	case "boolean":
		b, ok := value.(bool)
		if !ok {
			return "", fmt.Errorf("%w: parameter %q is declared boolean but got %T", ErrUnmirrorableValue, where, value)
		}
		return strconv.FormatBool(b), nil
	case "integer":
		return renderIntegerHeaderValue(where, value)
	default:
		// Unreachable: ParamHeaders admits only the three types above.
		return "", fmt.Errorf("%w: parameter %q has unsupported type %q", ErrUnmirrorableValue, where, h.Type)
	}
}

// renderIntegerHeaderValue renders an integer-declared parameter. JSON decoding
// yields float64 for every number, so an integer arrives as a float that must be
// checked for integrality and for SEP-2243's safe-integer range; the int/int64
// cases cover arguments built in Go rather than decoded from JSON.
func renderIntegerHeaderValue(where string, value any) (string, error) {
	switch n := value.(type) {
	case float64:
		if n != math.Trunc(n) {
			return "", fmt.Errorf("%w: parameter %q is declared integer but got %v", ErrUnmirrorableValue, where, n)
		}
		if n > maxSafeInteger || n < -maxSafeInteger {
			return "", fmt.Errorf(
				"%w: parameter %q value %v is outside the safe integer range", ErrUnmirrorableValue, where, n)
		}
		return strconv.FormatInt(int64(n), 10), nil
	case int:
		return renderIntegerHeaderValue(where, float64(n))
	case int64:
		return renderIntegerHeaderValue(where, float64(n))
	case json.Number:
		i, err := n.Int64()
		if err != nil {
			return "", fmt.Errorf("%w: parameter %q is declared integer but got %q", ErrUnmirrorableValue, where, n)
		}
		return renderIntegerHeaderValue(where, float64(i))
	default:
		return "", fmt.Errorf(
			"%w: parameter %q is declared integer but got %T", ErrUnmirrorableValue, where, value)
	}
}

// firstControlChar returns the first control character in s, reporting true when
// one exists. CR and LF are the header-injection vectors; NUL and the other C0
// controls are equally illegal in a header value.
func firstControlChar(s string) (string, bool) {
	for _, r := range s {
		if r < 0x20 || r == 0x7f {
			return string(r), true
		}
	}
	return "", false
}

// childPath returns path with seg appended, always in freshly allocated storage.
// Appending to path directly would let sibling recursions share (and overwrite)
// one backing array, so a path captured deeper in the walk could be rewritten by
// the next sibling. Allocating per child keeps every captured path independent.
func childPath(path []string, seg string) []string {
	child := make([]string, len(path)+1)
	copy(child, path)
	child[len(path)] = seg
	return child
}

// parseAnnotation validates a single x-mcp-header annotation against SEP-2243
// and records it in seen for the uniqueness check.
func parseAnnotation(raw any, node map[string]any, path []string, seen map[string]string) (ParamHeader, error) {
	where := strings.Join(path, ".")

	name, ok := raw.(string)
	if !ok {
		return ParamHeader{}, fmt.Errorf(
			"parameter %q: %s must be a string, got %T", where, XMCPHeaderAnnotation, raw)
	}
	if name == "" {
		return ParamHeader{}, fmt.Errorf("parameter %q: %s must not be empty", where, XMCPHeaderAnnotation)
	}
	if bad, invalid := firstNonTokenChar(name); invalid {
		return ParamHeader{}, fmt.Errorf(
			"parameter %q: %s value %q is not a valid HTTP field-name token (offending character %q)",
			where, XMCPHeaderAnnotation, name, bad)
	}

	// HTTP field names are case-insensitive, so two annotations differing only in
	// case would mirror onto the same header and one would silently win.
	folded := strings.ToLower(name)
	if first, dup := seen[folded]; dup {
		return ParamHeader{}, fmt.Errorf(
			"%s value %q on parameter %q collides case-insensitively with parameter %q",
			XMCPHeaderAnnotation, name, where, first)
	}
	seen[folded] = where

	declared, ok := node["type"].(string)
	if !ok {
		return ParamHeader{}, fmt.Errorf(
			"parameter %q: %s requires a declared primitive type (string, integer, or boolean)",
			where, XMCPHeaderAnnotation)
	}
	switch declared {
	case "string", "integer", "boolean":
	default:
		return ParamHeader{}, fmt.Errorf(
			"parameter %q: %s is not permitted on type %q (only string, integer, or boolean)",
			where, XMCPHeaderAnnotation, declared)
	}

	// childPath already gives each branch its own storage, so cloning is belt and
	// braces -- it keeps the returned Path independent of the walk even if the
	// traversal's path handling is later changed.
	return ParamHeader{Path: slices.Clone(path), Name: name, Type: declared}, nil
}

// firstNonTokenChar returns the first byte of s that is not an RFC 9110 tchar,
// reporting true when one exists. Operating on bytes rather than runes is
// correct here: every tchar is ASCII, so any multi-byte rune is invalid and its
// leading byte is a faithful thing to name in the error.
func firstNonTokenChar(s string) (string, bool) {
	for i := 0; i < len(s); i++ {
		if !isTokenChar(s[i]) {
			return string(s[i]), true
		}
	}
	return "", false
}

// isTokenChar reports whether c is an RFC 9110 tchar, the character set HTTP
// field names are drawn from.
func isTokenChar(c byte) bool {
	switch {
	case c >= 'a' && c <= 'z',
		c >= 'A' && c <= 'Z',
		c >= '0' && c <= '9':
		return true
	}
	switch c {
	case '!', '#', '$', '%', '&', '\'', '*', '+', '-', '.', '^', '_', '`', '|', '~':
		return true
	}
	return false
}
